drivers/at_cmd/at_cmd.c

```diff
@@ -48,6 +54,133 @@ struct callback_work_item {
+static notif_handler_t *find_notif_handler(at_cmd_handler_t handler)
+{
+	notif_handler_t *l_notif_handlers = notif_handlers;
+	notif_handler_t *ret = NULL;
+
+	/* Do nothing on empty list. */
+	if (l_notif_handlers == NULL) {
+		LOG_DBG("Empty list.\n");
+		return NULL;
+	}
+
+	/* Traverse through the list. */
+	k_sem_take(&cmd_pending, K_FOREVER);
+	do {
+		if (l_notif_handlers->handler == handler) {
+			ret = l_notif_handlers;
+			break;
+		} else {
+			l_notif_handlers = l_notif_handlers->next;
+		}
+	} while (l_notif_handlers != NULL);
+	k_sem_give(&cmd_pending);
+
+	return ret;
+}
+
+/**@brief Append notification handler. */
+static int append_notif_handler(at_cmd_handler_t handler)
+{
+	notif_handler_t *l_notif_handlers;
+	notif_handler_t *l_to_insert;
+
+	/* Check if it exists */
+	if (NULL != find_notif_handler(handler)) {
+		LOG_DBG("Already in the list.\n");
+		return 0;
+	}
+
+	k_sem_take(&cmd_pending, K_FOREVER);
+
+	/* Find the last element */
+	l_notif_handlers = notif_handlers;
+	if (l_notif_handlers != NULL) {
+		LOG_DBG("At least one element in the list.\n");
+		while (l_notif_handlers->next != NULL) {
+			l_notif_handlers = l_notif_handlers->next;
+		}
+	}
+
+	/* Allocate memory and fill. */
+	l_to_insert = (notif_handler_t *)k_malloc(sizeof(notif_handler_t));
+	if (l_to_insert == NULL) {
+		return -ENOBUFS;
+	}
+	l_to_insert->handler = handler;
+	l_to_insert->next    = NULL;
+
+	/* Append. */
+	if (l_notif_handlers != NULL) {
+		l_notif_handlers->next = l_to_insert;
+	} else {
+		notif_handlers = l_to_insert;
+	}
+	k_sem_give(&cmd_pending);
+
+	return 0;
+}
+
+/**@brief Remove notification handler. */
+static void remove_notif_handler(at_cmd_handler_t handler)
+{
+	notif_handler_t *l_notif_handlers = notif_handlers;
+	notif_handler_t *l_previous = NULL;
+
+	/* Do nothing on empty list. */
+	if (l_notif_handlers == NULL) {
+		LOG_DBG("Empty list.\n");
+		return;
+	}
+
+	/* 1st element. */
+	if (l_notif_handlers->handler == handler) {
+		LOG_DBG("Found 0x%08X as the 1st element.\n", (u32_t)handler);
+		k_sem_take(&cmd_pending, K_FOREVER);
+		notif_handlers = l_notif_handlers->next;
+		k_sem_give(&cmd_pending);
+
+		k_free(l_notif_handlers);
+		return;
+	}
+
+	/* Traverse through the list. */
+	k_sem_take(&cmd_pending, K_FOREVER);
+	do {
+		l_previous = l_notif_handlers;
+		l_notif_handlers = l_notif_handlers->next;
+
+		if (l_notif_handlers->handler == handler) {
+			LOG_DBG("Found 0x%08X elsewhere.\n", (u32_t)handler);
+			l_previous->next = l_notif_handlers->next;
+			k_free(l_notif_handlers);
+			break;
+		} 
+	} while (l_notif_handlers->next != NULL);
+	k_sem_give(&cmd_pending);
+}
+
+/**@brief Dispatcher for notification handlers. */
+static void notification_handler (char *response)
+{
+	notif_handler_t *l_notif_handlers = notif_handlers;
+
+	/* Do nothing on empty list. */
+	if (l_notif_handlers == NULL) {
+		return;
+	}
+
+	/* Go through each element in the list. */
+	k_sem_take(&cmd_pending, K_FOREVER);
```

Comment:
Reusing the `cmd_pending` semaphore for protecting the linked list might result in a deadlock (in case the command was waiting for a response while notification arrived). We should use a different semaphore (or mutex preferably) to protect the list. At this point, commands and notifications should not interfere with each other, so it should be safe to have a separate synchronization primitive for notifications.

Reply:
The adding and removing of nodes to the master list (handler_list) are now divided into two phases:
1. Queue to temporary lists handler_to_add and handler_to_remove. It can add a node to handler_to_add, so that the node will be added to handler_list, or it can add a node to handler_to_remove, so that the corresponding node in handler_list will be removed.
2. When the handlers from handler_list are not running or have finished running, the handler_list is then maintained according to the nodes stored in handler_to_add and handler_to_remove. This is done in function maintain_list(). Semaphore list_pending has been created for such purpose.

This way, deadlock does not happen when you try to call at_cmd_set/remove_notification_handler() from the handlers themselves.

Reply:
While I understand your motivation, I can't shake the feeling that we're highly complicating the code for a little benefit. 

Sure, giving extra flexibility for using notification_handler API is nice, but I'm not aware of an existing use case for this (@lemrey @evenl perhaps you know about something?). But now, instead of a single list, we end up in a quite complex design, with three lists and non-trivial to understand (and protect) synchronization mechanism. 

I'm really not in favour of introducing such complexity, especially that all I've asked for was a separate mutext/semaphore to protect list operations. But that's only my opinion, let's hear from other reviewers.